diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2a51382d..57cc8dec 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -16,7 +16,7 @@ if(${CMAKE_VERSION} VERSION_GREATER_EQUAL "3.16")
     if(APPLE)
         project(LLGL LANGUAGES C CXX OBJC OBJCXX)
     elseif(MSVC)
-        project(LLGL LANGUAGES C CXX CSharp)
+        project(LLGL LANGUAGES C CXX)
     else()
         project(LLGL LANGUAGES C CXX)
     endif()
